zebra-chain/src/value_balance.rs

```diff
@@ -0,0 +1,244 @@
+        })
+    }
+}
+
+impl<C> std::ops::Sub<ValueBalance<C>> for Result<ValueBalance<C>, ValueBalanceError>
+where
+    C: Constraint,
+{
+    type Output = Result<ValueBalance<C>, ValueBalanceError>;
+
+    fn sub(self, rhs: ValueBalance<C>) -> Self::Output {
+        let value_balance = self?;
+
+        Ok(ValueBalance::<C> {
+            transparent: (value_balance.transparent - rhs.transparent)?,
+            sprout: (value_balance.sprout - rhs.sprout)?,
+            sapling: (value_balance.sapling - rhs.sapling)?,
+            orchard: (value_balance.orchard - rhs.orchard)?,
+        })
+    }
+}
+
+#[cfg(any(test, feature = "proptest-impl"))]
+use proptest::prelude::*;
+#[cfg(any(test, feature = "proptest-impl"))]
+impl Arbitrary for ValueBalance<NegativeAllowed> {
+    type Parameters = ();
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        (
+            any::<Amount>(),
+            any::<Amount>(),
+            any::<Amount>(),
+            any::<Amount>(),
+        )
+            .prop_map(|(transparent, sprout, sapling, orchard)| Self {
+                transparent,
+                sprout,
+                sapling,
+                orchard,
+            })
+            .boxed()
+    }
+
+    type Strategy = BoxedStrategy<Self>;
+}
+
+#[cfg(any(test, feature = "proptest-impl"))]
+impl Arbitrary for ValueBalance<NonNegative> {
+    type Parameters = ();
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        (
+            any::<Amount<NonNegative>>(),
+            any::<Amount<NonNegative>>(),
+            any::<Amount<NonNegative>>(),
+            any::<Amount<NonNegative>>(),
+        )
+            .prop_map(|(transparent, sprout, sapling, orchard)| Self {
+                transparent,
+                sprout,
+                sapling,
+                orchard,
+            })
+            .boxed()
+    }
+
+    type Strategy = BoxedStrategy<Self>;
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    proptest! {
+        #[test]
+        fn test_operations(
+            value_balance1 in any::<ValueBalance>(),
+            value_balance2 in any::<ValueBalance>())
+        {
+            zebra_test::init();
+
+            // test the add operator
+            let add = Ok(value_balance1) + value_balance2;
+
+            if !add.is_err() {
+                prop_assert_eq!(add?, ValueBalance {
+                    transparent: (value_balance1.transparent + value_balance2.transparent)?,
+                    sprout: (value_balance1.sprout + value_balance2.sprout)?,
+                    sapling: (value_balance1.sapling + value_balance2.sapling)?,
+                    orchard: (value_balance1.orchard + value_balance2.orchard)?,
+                });
+            }
+            else {
+                let error_string = format!("{:?}", add.err().unwrap());
+                prop_assert_eq!(error_string.contains("AmountError"), true);
```

Comment:
Need some more work in these.

Reply:
I made a suggestion in the code to use `matches!` and the enum variant.